### ISSUE TITLE
List instead of set for SMEMBERS in 5.1

### ISSUE THREAD

**Issue by jules-ch** [redis/redis-py#3390]
**Version**: What redis-py and what redis version is the issue happening on?

**Platform**: What platform / version? Python 3.11 Ubuntu 24.04

**Description**: `smembers` is returning a list instead of a set in redis-py `5.1.0` which is an unintended breaking change.
Before:

```python
>>> from redis import Redis
>>> pool = Redis.from_url("redis://", decode_responses=True)
>>> pool.sadd("users", "user1")
0
>>> pool.smembers("users")
{'user1'}
```
AFter

```python
>>> from redis import Redis
>>> pool = Redis.from_url("redis://", decode_responses=True)
>>> pool.sadd("users", "user1")
0
>>> pool.smembers("users")
['user1']

**Comment by jules-ch** [comment]
Seems to be done in https://github.com/redis/redis-py/pull/3324#issue-2419053541

Can you add this in the list of the breaking changes in 5.1, it's not documented.
Also stubs from `types-redis need to be changed`

**Comment by dmaier-redislabs** [comment]
Hi Jules,

we had a note under bug fixes. I copied it over to the breaking changes section to make it more visible. Which RESP version are you using?

@vladvildanov I think we need to take a second look at this change.

Regards,
David

**Comment by jules-ch** [comment]
I did not specified RESP protocol version somewhere, is it tied to Redis server version ?

**Comment by vladvildanov** [comment]
@jules-ch Thanks for reaching us! Changes in those PR should only affect RESP3 users, I'll have a look on it

**Comment by dmaier-redislabs** [comment]
Hi @jules-ch ,

we did take another look and this impacts RESP2 and RESP3 with and without `hiredis-py` for response parsing. This was introduced as part of a bug fix in the context of dealing with RESP3 sets (message format). 

So we decided that the original fix should never have impacted the result of the set (data structure) commands. We will release a patch version of 5.1 that reverts this breaking change for those commands. The original fix will still have an intended impact on some commands (e.g., the `COMMAND` command), but we will ensure that all the `S*` commands (e.g., `SADD`) will behave semantically correct again by returning a set instead of a list. We apologize for the inconvenience that this has caused.

Regards,
David

**Comment by jules-ch** [comment]
@dmaier-redislabs Alright makes sense, thanks for the quick fix, I'll wait before updating.

Closing this issue.

**Comment by dmaier-redislabs** [comment]
Reopening it until we released 5.1.1

**Comment by aosh-ab7e** [comment]
The type difference was intentionally introduced by hiredis for parsing (possible) nested set. https://github.com/redis/hiredis-py/pull/189

Though I have never met the actual instance of a nested set, is it really OK to simply cast the response to set? 
(note: I agree with the fix as a user of redis-py, and reported the difference to hiredis https://github.com/redis/hiredis-py/issues/197)
